File: MatrixCode_v8/js/simulation/SimulationSystem.js
class SimulationSystem {
    constructor(grid, config) {
        this.grid = grid;
        this.config = config;
        this.streamManager = new StreamManager(grid, config);
        this.overlapInitialized = false;
        this._lastOverlapDensity = null;
        this.timeScale = 1.0;
    }

    update(frame) {
        this.streamManager.update(frame, this.timeScale);
        this._manageOverlapGrid(frame);
        this._updateCells(frame);
        
        if (this.grid.cellLocks) {
            this.grid.cellLocks.fill(0);
        }
    }

    _manageOverlapGrid(frame) {
        const s = this.config.state;

        if (!s.overlapEnabled) {
            if (this.overlapInitialized) {
                this.overlapInitialized = false;
                if (this.grid.secondaryChars && typeof this.grid.secondaryChars.fill === 'function') {
                    this.grid.secondaryChars.fill(32); 
                }
            }
            return;
        }
        
        const activeFonts = this.config.derived.activeFonts;
        const numFonts = activeFonts.length;
        const currentDensity = s.overlapDensity;
        
        const ovRgb = Utils.hexToRgb(s.overlapColor);
        const ovColor = Utils.packAbgr(ovRgb.r, ovRgb.g, ovRgb.b);

        const setOverlapChar = (i) => {
            let fIdx;
            if (this.grid.types[i] === CELL_TYPE.EMPTY) {
                fIdx = Math.floor(Math.random() * numFonts);
            } else {
                fIdx = this.grid.fontIndices[i];
            }
            
            const fontData = activeFonts[fIdx] || activeFonts[0];
            const chars = fontData.chars;
            let code = 32;
            if (chars && chars.length > 0) {
                const r = Math.floor(Math.random() * chars.length);
                code = chars[r].charCodeAt(0);
            }
            
            this.grid.secondaryChars[i] = code;
            this.grid.secondaryColors[i] = ovColor;
        };

        if (!this.overlapInitialized || this._lastOverlapDensity !== currentDensity) {
            const N = this.grid.secondaryChars.length;
            for (let i = 0; i < N; i++) {
                if (Math.random() < currentDensity) {
                    setOverlapChar(i);
                } else {
                    this.grid.secondaryChars[i] = 32; 
                }
            }
            this.overlapInitialized = true;
            this._lastOverlapDensity = currentDensity;
        }
    }

    _updateCells(frame) {
        const s = this.config.state;
        const d = this.config.derived;
        const grid = this.grid;

        for (const idx of grid.activeIndices) {
            this._updateCell(idx, frame, s, d);
        }
    }

    _updateCell(idx, frame, s, d) {
        const grid = this.grid;

        if (grid.cellLocks && grid.cellLocks[idx] === 1) return;

        const decay = grid.decays[idx];
        if (decay === 0) return;

        let age = grid.ages[idx];
        if (age > 0) {
            age = this._incrementAge(age, d.maxState);
            grid.ages[idx] = age;
        }

        // --- TRACER COLOR FADE ---
        // Transitions from Tracer Color -> Stream Color based on Age
        if (grid.types[idx] === CELL_TYPE.TRACER || grid.types[idx] === CELL_TYPE.ROTATOR) {
            const attack = s.tracerAttackFrames;
            const hold = s.tracerHoldFrames;
            const release = s.tracerReleaseFrames;
            const tracerColor = d.tracerColorUint32;
            const baseColor = grid.baseColors[idx];

            let ratio = 0; // 0 = Tracer, 1 = Base
            
            // Age 1..Attack: Fade In (Alpha handles this, color stays Tracer)
            // Attack..Attack+Hold: Hold Tracer Color
            // Attack+Hold..End: Fade to Stream Color
            
            const activeAge = age - 1;
            if (activeAge > attack + hold) {
                if (release > 0) {
                    ratio = Math.min(1.0, (activeAge - (attack + hold)) / release);
                } else {
                    ratio = 1.0;
                }
            }
            
            if (ratio >= 1.0) {
                grid.colors[idx] = baseColor;
                grid.glows[idx] = 0; // Remove glow after transition
            } else if (ratio > 0) {
                // Blend
                const tR = tracerColor & 0xFF;
                const tG = (tracerColor >> 8) & 0xFF;
                const tB = (tracerColor >> 16) & 0xFF;
                
                const bR = baseColor & 0xFF;
                const bG = (baseColor >> 8) & 0xFF;
                const bB = (baseColor >> 16) & 0xFF;
                
                const mR = Math.floor(tR + (bR - tR) * ratio);
                const mG = Math.floor(tG + (bG - tG) * ratio);
                const mB = Math.floor(tB + (bB - tB) * ratio);
                
                grid.colors[idx] = Utils.packAbgr(mR, mG, mB);
                grid.glows[idx] = s.tracerGlow * (1.0 - ratio);
            } else {
                // Hold Tracer
                grid.colors[idx] = tracerColor;
                grid.glows[idx] = s.tracerGlow;
            }
        }

        // Handle Rotator
        if (s.rotatorEnabled && grid.types[idx] === CELL_TYPE.ROTATOR) {
            this._handleRotator(idx, frame, s, d);
        }

        // Handle Dynamic Colors (Effects)
        if (grid.complexStyles.has(idx)) {
            const style = grid.complexStyles.get(idx);
            if (style.cycle) {
                const newHue = (style.h + style.speed) % 360;
                style.h = newHue; 
                const rgb = Utils.hslToRgb(newHue, style.s, style.l);
                grid.colors[idx] = Utils.packAbgr(rgb.r, rgb.g, rgb.b);
            }
        }

        // Handle Decay / Alpha
        if (decay >= 2) {
            // Ensure trails are Stream Color, not Tracer Color
            if (decay === 2) { // First frame of decay
                grid.colors[idx] = grid.baseColors[idx];
                grid.glows[idx] = 0;
            } else {
                // Also enforce it in case we missed frame 2 (unlikely but safe)
                grid.colors[idx] = grid.baseColors[idx];
            }
            
            grid.decays[idx]++;
            const newDecay = grid.decays[idx];
            if (this._shouldDecay(idx, newDecay, s.decayFadeDurationFrames)) {
                grid.clearCell(idx);
                return;
            }
            grid.alphas[idx] = this._calculateAlpha(idx, age, newDecay, s.decayFadeDurationFrames);
        } else {
            grid.alphas[idx] = this._calculateAlpha(idx, age, decay, s.decayFadeDurationFrames);
        }
    }

    _incrementAge(age, maxState) {
        return age + 1;
    }

    _handleRotator(idx, frame, s, d) {
        const grid = this.grid;
        const mix = grid.mix[idx]; 
        const decay = grid.decays[idx];

        if (mix > 0) {
            this._progressRotator(idx, mix, s.rotatorCrossfadeFrames);
        } else if (decay === 1 || (s.rotateDuringFade && decay > 1)) {
            this._cycleRotator(idx, frame, s.rotatorCrossfadeFrames, d.rotatorCycleFrames, s);
        }
    }

    _progressRotator(idx, currentMix, crossfadeFrames) {
        const grid = this.grid;
        const step = 1.0 / Math.max(1, crossfadeFrames);
        const newMix = currentMix + step;

        if (newMix >= 1.0) {
            const target = grid.getRotatorTarget(idx, false); 
            if (target) {
                grid.chars[idx] = target;
                if (this.config.state.overlapEnabled) {
                    const ovTarget = grid.getRotatorTarget(idx, true);
                    if (ovTarget) {
                        grid.secondaryChars[idx] = ovTarget;
                    }
                }
            }
            grid.mix[idx] = 0;
            grid.nextChars.delete(idx);
            grid.nextOverlapChars.delete(idx);
        } else {
            grid.mix[idx] = newMix;
        }
    }

    _cycleRotator(idx, frame, crossfadeFrames, cycleFrames, s) {
        const grid = this.grid;
        let effectiveCycle = cycleFrames;
        
        if (s.rotatorDesyncEnabled) {
            const variancePercent = s.rotatorDesyncVariance / 100;
            const maxVariance = cycleFrames * variancePercent;
            const offsetNorm = (grid.rotatorOffsets[idx] / 127.5) - 1.0;
            effectiveCycle = Math.max(1, Math.round(cycleFrames + (offsetNorm * maxVariance)));
        }

        if (frame % effectiveCycle === 0) {
            const fontIdx = grid.fontIndices[idx];
            const activeFonts = this.config.derived.activeFonts;
            const fontData = activeFonts[fontIdx] || activeFonts[0];
            const charSet = fontData.chars;
            
            const nextChar = this._getUniqueChar(grid.getChar(idx), charSet);
            const nextCode = nextChar.charCodeAt(0);
            
            let nextOvCode = 0;
            if (this.config.state.overlapEnabled) {
                const curOv = String.fromCharCode(grid.secondaryChars[idx]);
                const nextOv = this._getUniqueChar(curOv, charSet);
                nextOvCode = nextOv.charCodeAt(0);
            }

            if (crossfadeFrames <= 1) {
                grid.chars[idx] = nextCode;
                if (nextOvCode) grid.secondaryChars[idx] = nextOvCode;
            } else {
                grid.mix[idx] = 0.01; 
                grid.setRotatorTarget(idx, nextChar, false);
                if (nextOvCode) {
                    grid.setRotatorTarget(idx, String.fromCharCode(nextOvCode), true);
                }
            }
        }
    }
    
    _getUniqueChar(exclude, charSet) {
        if (!charSet) charSet = Utils.CHARS;
        if (charSet.length <= 1) return charSet[0];
        let char;
        let attempts = 0;
        do {
            char = charSet[Math.floor(Math.random() * charSet.length)];
            attempts++;
        } while (char === exclude && attempts < 10);
        return char;
    }

    _shouldDecay(idx, decay, fadeDurationFrames) {
        return decay > fadeDurationFrames + 2;
    }

    _calculateAlpha(idx, age, decay, fadeDurationFrames) {
        const b = this.grid.brightness[idx];
        if (age > 0 || decay === 1) {
            return 0.95 * b;
        } else if (decay >= 2) {
            const ratio = (decay - 2) / fadeDurationFrames;
            return 0.95 * (1 - ratio) * b;
        }
        return 0;
    }
}

File: C++Mac/SimulationSystem.cpp
#include "SimulationSystem.hpp"
#include "Utils.hpp"
#include <algorithm>
#include <cmath>

namespace Matrix {

SimulationSystem::SimulationSystem(CellGrid& g, Configuration& c, DerivedConfig& d)
    : grid(g), config(c), dConfig(d), streamManager(g, c, d) {
    
    rotatorSpeedMap.assign(60, 1.0f);
    for (int i = 0; i < 60; ++i) {
        rotatorSpeedMap[i] = 0.5f + Utils::RandomFloat(0.0f, 2.5f);
    }
}

SimulationSystem::~SimulationSystem() = default;

void SimulationSystem::Update(int frame) {
    if (config.simulationPaused) return;

    if (timeScale > 0.0f) {
        streamManager.Update(frame, timeScale);
        UpdateCells(frame, timeScale);
    } else if (timeScale < 0.0f) {
        streamManager.Update(frame, timeScale);
    }
}

void SimulationSystem::UpdateCells(int frame, float ts) {
    // Slow motion probabilistic update
    if (ts < 1.0f && Utils::RandomFloat(0.0f, 1.0f) > ts) return;

    // Use a copy of activeIndices to avoid modification during iteration
    std::vector<int> currentActive(grid.activeIndices.begin(), grid.activeIndices.end());
    for (int idx : currentActive) {
        UpdateCell(idx, frame, config, dConfig);
    }
}

void SimulationSystem::UpdateCell(int idx, int frame, const Configuration& s, const DerivedConfig& d) {
    uint8_t decay = grid.decays[idx];
    if (decay == 0) return;

    // Increment age
    int age = grid.ages[idx];
    if (age > 0) {
        age++;
        grid.ages[idx] = age;
    }

    uint8_t type = grid.types[idx];
    uint8_t baseType = type & CELL_TYPE_MASK;
    
    // Tracer Color Fade
    if (decay < 2 && (baseType == static_cast<uint8_t>(CellType::TRACER) || baseType == static_cast<uint8_t>(CellType::ROTATOR))) {
        int attack = s.tracerAttackFrames;
        int hold = s.tracerHoldFrames;
        int release = s.tracerReleaseFrames;
        
        uint32_t tracerColor = d.tracerColorUint32;
        uint32_t baseColor = grid.baseColors[idx];
        float ratio = 0.0f;
        int activeAge = age - 1;

        if (activeAge > (attack + hold)) {
            if (release > 0) {
                ratio = std::min(1.0f, static_cast<float>(activeAge - (attack + hold)) / static_cast<float>(release));
            } else {
                ratio = 1.0f;
            }
        }

        if (ratio >= 1.0f) {
            grid.colors[idx] = baseColor;
            grid.glows[idx] = 0.0f;
        } else if (ratio > 0.0f) {
            Color t = Color::UnpackABGR(tracerColor);
            Color b = Color::UnpackABGR(baseColor);
            uint8_t mR = static_cast<uint8_t>(t.r + (b.r - t.r) * ratio);
            uint8_t mG = static_cast<uint8_t>(t.g + (b.g - t.g) * ratio);
            uint8_t mB = static_cast<uint8_t>(t.b + (b.b - t.b) * ratio);
            grid.colors[idx] = Color::PackABGR(mR, mG, mB);
            grid.glows[idx] = s.tracerGlow * (1.0f - ratio);
        } else {
            grid.colors[idx] = tracerColor;
            grid.glows[idx] = s.tracerGlow;
        }
    }

    // Rotator
    if (baseType == static_cast<uint8_t>(CellType::ROTATOR)) {
        HandleRotator(idx, frame, s, d);
    }

    // Decay / Alpha
    if (decay >= 2) {
        if (decay == 2) {
            grid.colors[idx] = grid.baseColors[idx];
            grid.glows[idx] = 0.0f;
        }
        
        grid.decays[idx]++;
        int newDecay = grid.decays[idx];
        int maxFade = (grid.maxDecays[idx] > 0) ? grid.maxDecays[idx] : s.decayFadeDurationFrames;
        
        if (newDecay > (maxFade + 2)) {
            grid.ClearCell(idx);
            return;
        }
        
        grid.alphas[idx] = CalculateAlpha(idx, age, newDecay, maxFade);
    } else {
        int maxFade = (grid.maxDecays[idx] > 0) ? grid.maxDecays[idx] : s.decayFadeDurationFrames;
        grid.alphas[idx] = CalculateAlpha(idx, age, decay, maxFade);
    }
}

void SimulationSystem::HandleRotator(int idx, int frame, const Configuration& s, const DerivedConfig& d) {
    float mix = grid.mix[idx];
    if (mix > 0.0f) {
        ProgressRotator(idx, mix, s.rotatorCrossfadeFrames);
    } else if (s.rotatorEnabled) {
        CycleRotator(idx, frame, s.rotatorCrossfadeFrames, 10, s); // Simplified cycle duration
    }
}

void SimulationSystem::ProgressRotator(int idx, float currentMix, int crossfadeFrames) {
    float step = 1.0f / static_cast<float>(std::max(1, crossfadeFrames));
    float newMix = currentMix + step;

    if (newMix >= 1.0f) {
        uint16_t targetCode = grid.nextChars[idx];
        if (targetCode > 0) {
            grid.chars[idx] = targetCode;
        }
        grid.mix[idx] = 0.0f;
        grid.nextChars[idx] = 0;
    } else {
        grid.mix[idx] = newMix;
    }
}

void SimulationSystem::CycleRotator(int idx, int frame, int crossfadeFrames, int cycleFrames, const Configuration& s) {
    if (frame % cycleFrames == 0) {
        uint16_t nextCode = GetUniqueChar(grid.chars[idx]);
        if (crossfadeFrames <= 1) {
            grid.chars[idx] = nextCode;
        } else {
            grid.mix[idx] = 0.01f;
            grid.nextChars[idx] = nextCode;
        }
    }
}

uint16_t SimulationSystem::GetUniqueChar(uint16_t exclude) {
    std::string fallback = "0123456789Z:<=>\"*+-._!|";
    uint16_t c;
    do {
        c = static_cast<uint16_t>(fallback[Utils::RandomInt(0, fallback.length() - 1)]);
    } while (c == exclude);
    return c;
}

float SimulationSystem::CalculateAlpha(int idx, int age, int decay, int fadeDurationFrames) {
    float maxA = config.maxAlpha;
    float b = grid.brightness[idx];

    if (decay >= 2) {
        float ratio = static_cast<float>(decay - 2) / static_cast<float>(fadeDurationFrames);
        float fade = std::pow(std::max(0.0f, 1.0f - ratio), 2.0f);
        return maxA * fade * b;
    }

    int attack = config.tracerAttackFrames;
    if (age <= attack && attack > 0) {
        return maxA * (static_cast<float>(age) / static_cast<float>(attack)) * b;
    }

    return maxA * b;
}

} // namespace Matrix
